catroid/src/main/java/org/catrobat/catroid/content/bricks/ParameterizedBrick.kt

```diff
@@ -0,0 +1,197 @@
+
+    fun addBrick(brick: Brick): Boolean = loopBricks.add(brick)
+
+    override fun updateSelectorText() {
+        super.updateSelectorText()
+        val listSize = userLists.size
+        val firstLabel = view.findViewById<TextView>(R.id.brick_param_first_label)
+        firstLabel.text = view.resources.getQuantityString(
+            R.plurals.brick_parameterized_foreach_plural,
+            listSize, listSize
+        )
+        val secondLabel = view.findViewById<TextView>(R.id.brick_param_second_label)
+        secondLabel.text = view.resources.getQuantityString(
+            R.plurals.brick_parameterized_stored_plural,
+            listSize, listSize
+        )
+    }
+
+    override fun setCommentedOut(commentedOut: Boolean) {
+        super.setCommentedOut(commentedOut)
+        for (brick in loopBricks) {
+            brick.isCommentedOut = commentedOut
+        }
+        endBrick.isCommentedOut = commentedOut
+    }
+
+    @Throws(CloneNotSupportedException::class)
+    override fun clone(): Brick {
+        val clone = super.clone() as ParameterizedBrick
+        clone.endBrick = ParameterizedEndBrick(clone)
+        clone.loopBricks = mutableListOf()
+        clone.parameterizedData = ParameterizedData()
+        for (brick in loopBricks) {
+            clone.addBrick(brick.clone())
+        }
+        return clone
+    }
+
+    override fun consistsOfMultipleParts(): Boolean = true
+
+    override fun getAllParts(): List<Brick> {
+        val bricks = mutableListOf<Brick>()
+        bricks.add(this)
+        bricks.add(endBrick)
+        return bricks
+    }
+
+    override fun addToFlatList(bricks: MutableList<Brick>) {
+        super.addToFlatList(bricks)
+        for (brick in loopBricks) {
+            brick.addToFlatList(bricks)
+        }
+        bricks.add(endBrick)
+    }
+
+    override fun setParent(parent: Brick) {
+        super.setParent(parent)
+        for (brick in loopBricks) {
+            brick.parent = this
+        }
+        endBrick.parent = this
+    }
+
+    override fun getDragAndDropTargetList(): List<Brick> = loopBricks
+
+    override fun removeChild(brick: Brick): Boolean {
+        if (loopBricks.remove(brick)) {
+            return true
+        }
+
+        for (childBrick in loopBricks) {
+            if (childBrick.removeChild(brick)) {
+                return true
+            }
+        }
+        return false
+    }
+
+    override fun getViewResource(): Int = R.layout.brick_parameterized_input
+
+    override fun addActionToSequence(sprite: Sprite, sequence: ScriptSequenceAction) {
+        val repeatSequence =
+            ActionFactory.eventSequence(sequence.script) as ScriptSequenceAction
+        loopBricks.filterNot { brick -> brick.isCommentedOut }.forEach {
+            it.addActionToSequence(sprite, repeatSequence)
+        }
+        endBrick.addActionToSequence(sprite, repeatSequence)
+        parameterizedData?.reset()
```

Comment:
parameterizedData can't hold a null reference, ? isn't needed

Reply:
i like the ?. way more than the !!.

Reply:
I agree, what I meant was that "parameterizedData" is initialized as a non-null value so it can never be set to null - therefore this check is not needed. It doesn't really matter, just wanted to mention it.